util/wiki/magicWand.java

```diff
@@ -0,0 +1,113 @@
+  @*/
+  static Tree get_left(Tree t) {
+    //@ unfold t.tree();
+    Tree res = t.left;
+    /*@
+    create {
+      use Perm(t.value, write) ** Perm(t.priority, write) ** Perm(t.left, write) ** Perm(t.right, write);
+      use t.right != null ==> t.right.tree();
+      use res == t.left;
+      fold t.tree();
+      qed res.tree() -* t.tree();
+    }
+    @*/
+    return res;
+  }
+
+  /*@
+  context t != null;
+  context t.tree();
+  requires \unfolding t.tree() \in t.left != null;
+  @*/
+  static boolean check_left_priority(Tree t, int max_priority) {
+    Tree left = get_left(t);
+    // now we have tree(left) and a wand tree(left) -* tree(t)
+    //@ unfold left.tree();
+    boolean res = left.priority <= max_priority;
+    //@ fold left.tree();
+    //@ apply left.tree() -* t.tree();
+    // now left.tree() is no longer directly available, but t.tree() is back
+    return res;
+  }
+
+  /*@
+  requires t != null ** t.tree();
+  static boolean wand_helper(Tree t, int max_prio)
+    = \unfolding t.tree() \in t.priority <= max_prio;
+  @*/
+
+  /*@
+  yields int root_prio;
+  requires t != null;
+  requires t.tree() ** sorted(t);
+  requires \unfolding t.tree() \in t.left != null;
+  ensures \result != null ** \result.tree();
+  ensures sorted(\result) && wand_helper(\result, root_prio);
+  ensures \result.tree() ** sorted(\result) ** wand_helper(\result, root_prio) 
+          -* 
+          t.tree() ** sorted(t);
+  @*/
+  static Tree get_left_sorted(Tree t) {
+    //@ unfold t.tree();
+    //@ ghost root_prio = t.priority;
+    Tree res = t.left;
+    /*@
+    create {
+      // necessary parts for t.tree()
+      use Perm(t.value, write) ** Perm(t.priority, write) ** Perm(t.left, write) ** Perm(t.right, write);
+      use t.right != null ==> t.right.tree();
+      use res == t.left;
+      // necessary parts for t.sorted()
+      use res != null;
+      use t.priority == root_prio;
+      use t.right != null ==> sorted(t.right) && \unfolding t.right.tree() \in t.priority >= t.right.priority;
+      // assemble parts
+      fold t.tree();
+      qed res.tree() ** sorted(res) ** wand_helper(res, root_prio) 
+          -* 
+          t.tree() ** sorted(t);
+    }
+    @*/
+    return res;
+  }
+
+  /*@
+  context t != null;
+  context t.tree() ** sorted(t);
+  requires \unfolding t.tree() \in t.left != null;
+  @*/
+  static boolean check_left_priority_sorted(Tree t, int max_priority) {
+    //@ ghost int root_prio;
+    Tree left = get_left_sorted(t) /*@ then { root_prio = root_prio; } @*/;
+    // now we have left.tree(), sorted(left) and wand_helper(left, root_prio), as well as the wand
+    //@ unfold left.tree();
+    boolean res = left.priority <= max_priority;
+    //@ fold left.tree();
+    //@ apply (left.tree() ** sorted(left) ** wand_helper(left, root_prio)) -* (t.tree() ** sorted(t));
+    // now left.tree() is no longer directly available, but t.tree() is back and sorted
+    return res;
+  }
+}
```

Comment:
hmm?

Reply:
Removed it, or was there something that surprised you?